### PR TITLE
Investigating spurious windows errors

### PR DESCRIPTION
**Pull request by matklad**
r? @alexcrichton

### CODE REVIEW

File: tests/testsuite/build_script.rs [hunk 1]
```diff
@@ -3812,7 +3810,8 @@ fn rename_with_link_search_path() {
         &root.join("foo.dll.lib"),
         p2.root().join("foo.dll.lib"),
     ));
-    fs::remove_dir_all(p.root()).unwrap();
+    remove_dir_all(p.root())
```
Comment (matklad) [inline tests/testsuite/build_script.rs]: This `remove_dir_all` should at least tell us which file causes problems. It also has some special handling for `Permission Denied`, so I even have a slight hope that this'll just fix the test.

Comment (matklad) [reply tests/testsuite/build_script.rs]: To clarify, I am not sure which unwrap fails exactly, but looking at the output of https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.4613#L1554 and several similar failures, looks like it happens in `test` itself, and not in the cargo-under-test.

**Comment by ehuss** [comment]
I did a bunch of investigation yesterday.  I'm close to figuring out what's wrong.  I have reliable reproductions.  In general, it looks like it is problematic to immediately rename or delete a binary after executing it.  I wrote a test that simulates `changing_bin_features_caches_targets` by doing the following:

1. Link B to A
2. Run A
3. Del A
4. Link C to A
5. Run A
6. Del A

The link calls will sometimes fail with access denied.  It's like the call to unlink leaves a ghost behind.  I've even caught it where A.exists() is true immediately after a successful unlink!  

I have Defender disabled, indexing disabled, and it's not related to mspdbsrv (although all 3 of those can make it substantially worse).  I'd like to understand why there seems to be a handle on the file.  However, a quick workaround is to add a 1 second delay after executing a file before deleting it.

**Comment by ehuss** [comment]
BTW, I'm also able to reproduce the spurious error in `test_edition` on Linux, but I haven't looked into why it happens, yet.

**Comment by alexcrichton** [comment]
@bors: r+

Awesome thanks so much for looking into this @ehuss! If you've got the time mind opening an issue with your findings? 

In the meantime this seems good to land so I will do so

**Pull request merged**